### ISSUE TITLE
FR: User sign-up portal

### ISSUE THREAD

**Issue by yaleman** [kanidm/kanidm#2287]
**Is your feature request related to a problem? Please describe.**

For a bunch of deployment use cases it'd be good to have a user sign-up portal, or enable developers to integrate signup into another system.

**Describe the solution you'd like**

This'd be an optional UI path on Kanidm's frontend which allows users to register as a new `person` which is disabled until the admin enables it. I'd probably put them in a group like `new_signups` by default, too. 

The UI would prompt for:

* Preferred Username
* Display Name (optional?)
* Email Address (optional)
* Confirmation of acceptance of Terms of Service (with a link)

CSRF on the form is required as a minimum, but some kind of captcha makes sense from a rate limiting/human validation perspective to limit spam attacks.

I really don't want to suggest email validation, so the idea is basically "create the user, tell them that it's done, and that the admin will contact them with details soon".

**Describe alternatives you've considered**

Requiring everyone to manage their own thing with a separate service, service account and a coop full of chickens.

**Additional context**

Long-term, this should be something that can be enabled/disabled without a restart - in case of a spam attack you don't want to have to restart the server and interrupt auth to stop the flood.

I'd suggest doing it in raw HTML and form things because it's *very* simple, doesn't need fancy WASM/Javascript things and can be rendered server-side trivially.

```mermaid
graph TD;

S[Start] --GET--> F["Signup Form (/signup)"]
F --POST--> B["Form validation (/signup/submit)"]
B -. Failed Somehow, redirect with error .->F
B -. Form is valid .-> IDM["IDM Creates Person"]
IDM --HTTP Redirect--> T["Thanks Page (/signup/complete)"]
```

**Comment by Firstyear** [comment]
I actually think that potentially the best thing here is a separate service with it's own ui, and it has an api token that allows it to create users. That way it doesn't necessarily have to be built in, and other people can make their own portals.

**Comment by yaleman** [comment]
Yeah, I thought about that as an option too, I was mainly trying to avoid starting a whole extra service/app for it 🙄

**Comment by Firstyear** [comment]
I have lots of thoughts on this, I think making it separate is the "most flexible" if only because it allows people to expose it in a separate way to the rest of the Kani application. I can see a user sign up being something that people want to customise the most of anything, especially around captchas and things. And they may even want to integrate user sign up to an existing application that they have. This is why I was thinking a "toolbox" to allow someone to make a signup portal would be better because that way they can more easily integrate to their existing stack.

But then again, the big challenge here would be pw resets and credential changes since Kani currently controls those too. I wonder how we'd deal with that?

**Comment by yaleman** [comment]
A lot of those considerations are why I was considering an integrated MVP to test the waters on the whole thing - to avoid having to handle all the HTTPS /app startup and routing and all that stuff, which makes for a *lot* of work 😄

One of my original concept write-ups was a toolbox, and there's no reason it couldn't be a target, but if you consider smaller deployments where they're not building an enterprise platform to hide Kanidm, and you just want to go "hey, we use Kanidm and allow admin-gated signups" then it's still a feature worth building, IMO.

> But then again, the big challenge here would be pw resets and credential changes since Kani currently controls those too. I wonder how we'd deal with that?

If you were using an external service account to control it, you'd need to give it scoped permissions to create password resets (for that **specific** group) so it'd be able to send the link / QR code to the user, would be my guess. 

Or... at some point Kanidm's going to have to learn how to send emails (preferably with some kind of API-driven integrations like AWS SES / anything but integrating our own SMTP client 🤮)
